doc/dev/sharp_bits.rst

```diff
@@ -0,0 +1,1015 @@
+  ``jnp.where`` where possible.
+
+For more details, please see the `JAX documentation
+<https://jax.readthedocs.io/en/latest/notebooks/Common_Gotchas_in_JAX.html>`__.
+
+JAX integration
+---------------
+
+Compiled functions remain JAX compatible, and you can call JAX transformations
+on them, such as ``jax.grad`` and ``jax.vmap``. You can even call ``jax.jit``
+on functions that call qjit-compiled functions:
+
+>>> dev = qml.device("lightning.qubit", wires=2)
+>>> @qjit
+... @qml.qnode(dev)
+... def circuit(x):
+...     qml.RX(x, wires=0)
+...     return qml.expval(qml.PauliZ(0))
+>>> @jax.jit
+... def workflow(y):
+...     return jax.grad(circuit)(jnp.sin(y))
+>>> workflow(0.6)
+Array(-0.53511382, dtype=float64, weak_type=True)
+
+However, a ``jax.jit`` function calling a ``qjit`` function will always result
+in a callback to Python, so will be slower than if the function was purely compiled
+using ``jax.jit`` or ``qjit``.
+
+If you want to compile some functionality that is not currently Catalyst
+compatible, or you want to make use of JAX-supported hardware such as TPUs
+for classical processing, mixing ``jax.jit`` and ``qjit`` will allow this.
+However, if possible, try to always use ``qjit`` to compile your entire
+workflow.
+
+Internal QJIT transformations
+-----------------------------
+
+Inside of a qjit-compiled function, JAX transformations
+(``jax.grad``, ``jax.jacobian``, ``jax.vmap``, etc.)
+can be used **as long as they are not applied to quantum processing**.
+
+>>> @qjit
+... def f(x):
+...     def g(y):
+...         return -jnp.sin(y) ** 2
+...     return jax.grad(g)(x)
+>>> f(0.4)
+array(-0.71735609)
+
+If they are applied to quantum processing, an error will occur:
+
+>>> @qjit
+... def f(x):
+...     @qml.qnode(dev)
+...     def g(y):
+...         qml.RX(y, wires=0)
+...         return qml.expval(qml.PauliX(0))
+...     return jax.grad(lambda y: g(y) ** 2)(x)
+>>> f(0.4)
+NotImplementedError: must override
+
+Instead, only Catalyst transformations will work when applied to hybrid
+quantum-classical processing:
+
+>>> @qjit
+... def f(x):
+...     @qml.qnode(dev)
+...     def g(y):
+...         qml.RX(y, wires=0)
+...         return qml.expval(qml.PauliZ(0))
+...     return grad(lambda y: g(y) ** 2)(x)
+>>> f(0.4)
+array(-0.71735609)
+
+Always use the equivalent Catalyst transformation
+(:func:`catalyst.grad`, :func:`catalyst.jacobian`, :func:`catalyst.vjp`, :func:`catalyst.jvp`)
+inside of a qjit-compiled function.
+
+Inspecting and drawing circuits
+-------------------------------
+
+A useful tool for debugging quantum algorithms is the ability to draw them. Currently,
+:func:`@qjit <~.qjit>` compiled QNodes used as input to
+:func:`qml.draw <pennylane.draw>`, with the following caveats:
+
+- :func:`qml.draw <pennylane.draw>` call must occur outside the :func:`@qjit <.qjit>`
+
+- The :func:`@qjit <.qjit>` decorator must be placed directly on top of the QNode
```

Comment:
I think this is slightly misleading, you should be able to draw QNodes even if they don't have a qjit decorator, as in:
```py
@qml.qnode(dev)
def circuit(x):
    qml.Hadamard(0)
    return qml.expval(qml.PauliZ(0))
    
@qjit
def workflow(x):
    return circuit(x)
    
qml.draw(circuit)(x)
```
has always worked. This form was the recent addition:
 ```py
```py
@qjit
@qml.qnode(dev)
def circuit(x):
    qml.Hadamard(0)
    return qml.expval(qml.PauliZ(0))

qml.draw(circuit)(x)
```

Reply:
```suggestion
- If used, the :func:`@qjit <.qjit>` decorator must be placed directly on top of the QNode (that is, you cannot draw a hybrid `qjit` compiled function that _contains_ a QNode).
```

Reply:
I'm still not quite sure what the intended message is. Do you have an example for the situation that you are trying to exclude?

Reply:
It's maybe a bit obvious, but I'm trying to exclude two things:

```python
@qml.qnode(dev)
def circuit(x):
    ...

@qjit
def function(x):
    qml.draw(circuit)(x)
```

and 


```python
@qjit
def hybrid_function(x):
    @qml.qnode(dev)
    def circuit(x):
        ...

qml.draw(hybrid_function)(x)
```

Both are super obviously not going to work, but both tripped me up in the past.

Reply:
so I'm basically trying to say 'you can only draw circuits of the form `qjit(qnode)` or `qnode`; don't try and draw anything else

Reply:
Oh interesting, would the second one work in pennylane?

Reply:
Nope it doesn't work either 😆 Don't know why I kept trying to do it.